### PR TITLE
[ios][web-browser] Fix browser session living longer than VC

### LINKED ISSUE #22406: [expo-web-browser] audio don't stop playing after pressing "Done" button - iOS - Bare
### Minimal reproducible example

https://github.com/arekm213/expo-web-browser-audio-dont-stop-playing-repro

### Summary

Users using our app are reporting that audio from website opened by expo-web-browser don't stop playing after pressing "Done" button (which closes web browser modal) on **iOS** in **Bare** workflow, its not how its supposed to be, right? I tried to add `WebBrowser.dismissBrowser();` after `await WebBrowser.openBrowserAsync()` and it makes audio stop playing.
In our app we added patch below to prevent this behaviour:

```
diff --git a/node_modules/expo-web-browser/build/WebBrowser.js b/node_modules/expo-web-browser/build/WebBrowser.js
index 00a596d..b15c735 100644
--- a/node_modules/expo-web-browser/build/WebBrowser.js
+++ b/node_modules/expo-web-browser/build/WebBrowser.js
@@ -135,6 +135,9 @@ export async function openBrowserAsync(url, browserParams = {}) {
     finally {
         // WebBrowser session complete, unset lock
         browserLocked = false;
+        if(Platform.OS === 'ios') {
+            dismissBrowser();
+        }
     }
     return result;
 }
```
Do you think this is safe?

To repro you can just run the minimal reproducible example repo i attached and click orange button - this is gonna redirect you to a tweet, play the video and press "Done" button - audio from that website will still be playing

 

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 12.6
      Shell: 5.8.1 - /bin/zsh
    Binaries:
      Node: 18.15.0 - /opt/homebrew/bin/node
      Yarn: 1.22.18 - /opt/homebrew/bin/yarn
      npm: 9.5.0 - /opt/homebrew/bin/npm
      Watchman: 2023.03.27.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.11.3 - /Users/arekmrowczyk/.rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.1, iOS 16.1, macOS 13.0, tvOS 16.1, watchOS 9.1
    IDEs:
      Android Studio: 2021.3 AI-213.7172.25.2113.9014738
      Xcode: 14.1/14B47b - /usr/bin/xcodebuild
    npmPackages:
      expo: ~48.0.15 => 48.0.16 
      react: 18.2.0 => 18.2.0 
      react-native: 0.71.7 => 0.71.7 
    npmGlobalPackages:
      expo-cli: 6.3.2
    Expo Workflow: bare

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
Closes #22406

# How
Destroy the session after the `SFSafariViewController` has been dismissed

# Test Plan
Tested in bare expo

### CODE REVIEW

File: packages/expo-web-browser/ios/WebBrowserModule.swift [hunk 1]
```diff
@@ -15,14 +15,15 @@ final public class WebBrowserModule: Module {
       guard self.currentWebBrowserSession?.isOpen != true else {
         throw WebBrowserAlreadyOpenException()
       }
-      self.currentWebBrowserSession = WebBrowserSession(url: url, options: options)
+      self.currentWebBrowserSession = WebBrowserSession(url: url, options: options) {
+        self.destroyBrowserSession()
```
[the remark this answers is not in the document]
Comment (alanjhughes) [reply packages/expo-web-browser/ios/WebBrowserModule.swift:19]: Yeah we could do that and make the callback accept the type to resolve with.

File: packages/expo-web-browser/ios/WebBrowserModule.swift [hunk 2]
```diff
@@ -7,23 +7,28 @@ import AuthenticationServices
...
   private var currentAuthSession: WebAuthSession?
+  private var promise: Promise?
 
   public func definition() -> ModuleDefinition {
     Name("ExpoWebBrowser")
 
-    AsyncFunction("openBrowserAsync") { (url: URL, options: WebBrowserOptions, promise: Promise) throws in
-      guard self.currentWebBrowserSession?.isOpen != true else {
+    AsyncFunction("openBrowserAsync") { (url: URL, options: WebBrowserOptions, promise: Promise) in
+      guard self.promise == nil else {
```
Comment (tsapeta) [inline packages/expo-web-browser/ios/WebBrowserModule.swift:16]:
```suggestion
      guard self.currentWebBrowserSession == nil else {
```

File: packages/expo-web-browser/ios/WebBrowserModule.swift [hunk 3]
```diff
@@ -7,23 +7,28 @@ import AuthenticationServices
...
-      self.currentWebBrowserSession = WebBrowserSession(url: url, options: options) {
-        self.destroyBrowserSession()
+      self.currentWebBrowserSession = WebBrowserSession(url: url, options: options) { type in
+        self.promise?.resolve(["type": type])
+        self.promise = nil
+        self.currentWebBrowserSession = nil
       }
-      self.currentWebBrowserSession?.open(promise)
+      self.currentWebBrowserSession?.open()
+      self.promise = promise
```
Comment (tsapeta) [inline packages/expo-web-browser/ios/WebBrowserModule.swift:25]: We probably don't have to keep the promise here. The `onDismiss` closure can just capture the local `promise` 🤔

File: packages/expo-web-browser/ios/WebBrowserModule.swift [hunk 4]
```diff
@@ -7,23 +7,28 @@ import AuthenticationServices
...
-      guard self.currentWebBrowserSession?.isOpen != true else {
+    AsyncFunction("openBrowserAsync") { (url: URL, options: WebBrowserOptions, promise: Promise) in
+      guard self.promise == nil else {
         throw WebBrowserAlreadyOpenException()
       }
-      self.currentWebBrowserSession = WebBrowserSession(url: url, options: options) {
-        self.destroyBrowserSession()
+      self.currentWebBrowserSession = WebBrowserSession(url: url, options: options) { type in
+        self.promise?.resolve(["type": type])
+        self.promise = nil
```
Comment (tsapeta) [inline packages/expo-web-browser/ios/WebBrowserModule.swift:21]:
```suggestion
      self.currentWebBrowserSession = WebBrowserSession(url: url, options: options) { [promise] type in
        promise.resolve(["type": type])
```

**Review by tsapeta — APPROVED** [review]

**Pull request merged**